PolyEngine/Core/Src/RTTI/RTTI2.hpp

```diff
@@ -0,0 +1,208 @@
+                    }
+                });
+                id = std::type_index{typeid(T)};
+                listId = std::type_index{typeid(RTTIinfo<T>::info)};
+				if constexpr (std::is_default_constructible<T>::value)
+					construct = [](){ return new T; };
+				if constexpr (std::is_destructible<T>::value)
+					destroy = [](void* p) { delete (T*)p; };
+            }
+            template <typename T> const typename AttrType<T>::type::container * const get() const
+            { 
+                auto i = attrs.find(typeid(T));
+                return i != attrs.end() ? std::any_cast<typename AttrType<T>::type::container>(&i->second) : nullptr;
+            }
+        };
+        template <typename T> const TypeInfo& TypeManager::registerOrGetType()
+        {
+            const static TypeInfo& ti = [this](){
+                if(byId.count(typeid(T))) {
+                    TypeInfo& ti = *byId.at(typeid(T));
+                    // not sure if this is possible to happen
+                    if(ti.listId != std::type_index{typeid(RTTIinfo<T>::info)})
+                        throw std::runtime_error("attempting to register list with different list");
+                    return ti;
+                }
+                auto p = std::make_shared<TypeInfo>();
+                p->init<T>();
+                //insert into byName only if name exists
+                runList(RTTIinfo<T>::info, [&, this](auto x) {
+                    if constexpr (std::is_same<classname, decltype(x)>::value) {
+                        if(byName.count(x.name)) {
+                            std::string s = "different class with name ";
+                            s += x.name;
+                            s += " is alredy registered";
+                            throw std::runtime_error(s);
+                        }
+                        byName.insert_or_assign(x.name, p);
+                    }
+                });
+                byId.insert_or_assign(p->id, p);
+				return *p;
+            }();
+            return ti;
+        }
+
+        template <typename T> bool isSame(const TypeInfo& ti)
+        {
+            return std::type_index{typeid(T)} == ti.id;
+        }
+        
+        template <typename T> bool isDerived(const TypeInfo& ti)
+        {
+            return std::any_of(ti.bases.begin(), ti.bases.end(), [&](const TypeInfo& base) {
+                return base.id == std::type_index{typeid(T)} || isDerived<T>(base); //dfs
+            });
+        }
+
+        template <typename T> bool isDerivedOrSame(const TypeInfo& ti)
+        {
+            return isSame<T>(ti) || isDerived<T>(ti);
+        }
+
+        struct RTTIBase {
+            const TypeInfo& typeInfo;
+			RTTIBase() = delete;
+			RTTIBase(const TypeInfo& ti) : typeInfo(ti) {}
+            template <typename T> bool isOfType() { return isDerivedOrSame<T>(typeInfo); }
+			~RTTIBase() {}
+        };
+		
+		//for some reason this doesn't work, maybe someone will enlighten me why
```

Comment:
Template exporting through DLL-s? I'm not sure because I don't know what kind of issue do you have with it.

Reply:
basicly i wanted to avoid ceremony when deriving from RTTIBase, with deriving from this instead, but it kept trying to call default constructor of RTTIBase for some reason

Reply:
Then we need to have a better look at it. This behavior is indeed strange. We will discuss this further IRL.

Reply:
This couldn't work, deleted